drivers/flash/soc_flash_nrf.c

```diff
@@ -88,16 +88,18 @@ static inline bool is_aligned_32(u32_t data)
 	return (data & 0x3) ? false : true;
 }
 
-static inline bool is_regular_addr_valid(off_t addr, size_t len)
+static inline bool is_regular_addr_valid(off_t *addr, size_t len)
 {
 	size_t flash_size = nrfx_nvmc_flash_size_get();
 
-	if (addr >= DT_FLASH_BASE_ADDRESS + flash_size ||
-	    addr < DT_FLASH_BASE_ADDRESS ||
+	if (*addr >= flash_size ||
```

Comment:
This can be simplified to:
```if ((addr + len > flash_size) || ((addr + len) < addr)) {```

first part is the check, second part protection against wrap around.

Reply:
That would be true if all operands were unsigned, but `addr` is signed so the calculation `addr + len` may produce undefined behavior.  The following is the same idea but safe:

```
static inline bool adjusted_offset_valid(off_t *offs, size_t len)
{
  if (*offs < 0) { /* Skip if you trust the caller */
    return false;
  }

  /* Write to [start, stop) as a sub-region of [0, size). */
  size_t size = nrfx_nvmc_flash_size_get()
  size_t start = (size_t)*offs;
  size_t stop = start + len;

  if ((stop > size)        /* would write past size */
      || (stop < start)) { /* would wrap past size */
    return false;
  }

  *offs = start + (size_t)DT_FLASH_BASE_ADDRESS;
  return true;
}
```

Reply:
@pabigot, you are right, off_t is signed. I have always considered it unsigned. Should not all off_t be replaced with u32_t in the flash API ?

Reply:
Or `size_t`, but in my opinion: yes.  Actually every API should be reviewed and inappropriate use of signed types (e.g. for gpio flags, and intervals) replaced by unsigned types.  The problem is pervasive in Zephyr.